TestGetAllXattr failing due to llistxattr() returning different lengths

* Distribution: Debian
 * Distribution version: sid
 * LXD version: 5.0.0 LTS
 * Go version: 1.18.4

This issue was discovered by another individual while reviewing the Debian LXD packaging and is reproducible in Debian's CI pipeline, but I have not been able to reproduce it myself locally.

When the tests are run, `TestGetAllXattr` fails:

```
=== RUN   TestGetAllXattr
    util_linux_test.go:79: numerical result out of range
--- FAIL: TestGetAllXattr (0.00s)
```

Adding some additional debugging output, I can see that for whatever reason the initial reported length of the extended attributes for the directory created (but not the file) is unexpectedly 23 bytes longer than it really is which causes the buffer to be padded with NULLs at the end:

```
=== RUN   TestGetAllXattr
I got pre=37 and post=37
00000000  75 73 65 72 2e 63 68 65  63 6b 73 75 6d 00 75 73  |user.checksum.us|
00000010  65 72 2e 72 61 6e 64 6f  6d 00 75 73 65 72 2e 65  |er.random.user.e|
00000020  6d 70 74 79 00                                    |mpty.|
I split into 3 strings
xattr -> user.checksum
xattr -> user.random
xattr -> user.empty
I got pre=60 and post=37
00000000  75 73 65 72 2e 63 68 65  63 6b 73 75 6d 00 75 73  |user.checksum.us|
00000010  65 72 2e 65 6d 70 74 79  00 75 73 65 72 2e 72 61  |er.empty.user.ra|
00000020  6e 64 6f 6d 00 00 00 00  00 00 00 00 00 00 00 00  |ndom............|
00000030  00 00 00 00 00 00 00 00  00 00 00 00              |............|
I split into 26 strings
xattr -> user.checksum
xattr -> user.empty
xattr -> user.random
xattr -> 
    util_linux_test.go:79: numerical result out of range
--- FAIL: TestGetAllXattr (0.00s)
```

If I make a slice of the buffer to truncate the excess NULLs with the following one line change, then the test passes:

```
diff --git a/shared/util_linux.go b/shared/util_linux.go
index 24cbed542..cbdfd3384 100644
--- a/shared/util_linux.go
+++ b/shared/util_linux.go
@@ -115,7 +115,7 @@ func GetAllXattr(path string) (xattrs map[string]string, err error) {
                return nil, fmt.Errorf("Extended attribute list size increased from %d to %d during retrieval", pre, post)
        }
 
-       split := strings.Split(string(dest), "\x00")
+       split := strings.Split(string(dest[:post]), "\x00")
        if split == nil {
                return nil, fmt.Errorf("No valid extended attribute key found")
        }
```

```
=== RUN   TestGetAllXattr
I got pre=37 and post=37
00000000  75 73 65 72 2e 65 6d 70  74 79 00 75 73 65 72 2e  |user.empty.user.|
00000010  63 68 65 63 6b 73 75 6d  00 75 73 65 72 2e 72 61  |checksum.user.ra|
00000020  6e 64 6f 6d 00                                    |ndom.|
00000000  75 73 65 72 2e 65 6d 70  74 79 00 75 73 65 72 2e  |user.empty.user.|
00000010  63 68 65 63 6b 73 75 6d  00 75 73 65 72 2e 72 61  |checksum.user.ra|
00000020  6e 64 6f 6d 00                                    |ndom.|
I split into 3 strings
xattr -> user.empty
xattr -> user.checksum
xattr -> user.random
I got pre=60 and post=37
00000000  75 73 65 72 2e 65 6d 70  74 79 00 75 73 65 72 2e  |user.empty.user.|
00000010  72 61 6e 64 6f 6d 00 75  73 65 72 2e 63 68 65 63  |random.user.chec|
00000020  6b 73 75 6d 00 00 00 00  00 00 00 00 00 00 00 00  |ksum............|
00000030  00 00 00 00 00 00 00 00  00 00 00 00              |............|
00000000  75 73 65 72 2e 65 6d 70  74 79 00 75 73 65 72 2e  |user.empty.user.|
00000010  72 61 6e 64 6f 6d 00 75  73 65 72 2e 63 68 65 63  |random.user.chec|
00000020  6b 73 75 6d 00                                    |ksum.|
I split into 3 strings
xattr -> user.empty
xattr -> user.random
xattr -> user.checksum
--- PASS: TestGetAllXattr (0.00s)
```

However, I don't know if that's a safe thing to do because I don't understand what's causing the initial call to `llistxattr()` to return a longer length than the actual data.

I've tried to reproduce this locally, but haven't been able to do so. I have tried running the test directly on a physical machine, within a LXD container, and within a docker container. I have tried with `/tmp/` mounted on ext4, zfs, and overlay filesystems. When `/tmp/` is mounted with `noxattr` or as tmpfs, the test correctly short-circuits out and passes.

This is being seen in LXD 5.0.0 LTS; I've looked at the git history for `shared/util_linux.go` since that release and don't see anything that I think would affect this behavior. Since I can't reproduce this locally I haven't been able to try building the master branch of LXD to confirm that. If anyone is able to share steps to reproduce or give suggestions on what I could try, I'd appreciate it!

--- a/go.mod
+++ b/go.mod
@@ -102,6 +102,7 @@ require (
 	github.com/pelletier/go-toml/v2 v2.0.5 // indirect
 	github.com/philhofer/fwd v1.1.1 // indirect
 	github.com/pkg/errors v0.9.1 // indirect
+	github.com/pkg/xattr v0.4.8 // indirect
 	github.com/pmezard/go-difflib v1.0.0 // indirect
 	github.com/power-devops/perfstat v0.0.0-20220216144756-c35f1ee13d7c // indirect
 	github.com/prometheus/procfs v0.8.0 // indirect

--- a/shared/idmap/idmapset_linux.go
+++ b/shared/idmap/idmapset_linux.go
@@ -591,7 +591,7 @@ func (set *IdmapSet) doUidshiftIntoContainer(dir string, testmode bool, how stri
 	tmp := filepath.Dir(dir)
 	tmp, err := filepath.EvalSymlinks(tmp)
 	if err != nil {
-		return fmt.Errorf("Expand symlinks: %w", err)
+		return fmt.Errorf("Failed expanding symlinks of %q: %w", tmp, err)
 	}
 
 	dir = filepath.Join(tmp, filepath.Base(dir))
@@ -669,7 +669,7 @@ func (set *IdmapSet) doUidshiftIntoContainer(dir string, testmode bool, how stri
 					if how != "in" || atomic.LoadInt32(&VFS3Fscaps) == VFS3FscapsSupported {
 						err = SetCaps(path, caps, rootUid)
 						if err != nil {
-							logger.Warnf("Unable to set file capabilities on %s", path)
+							logger.Warnf("Unable to set file capabilities on %q: %v", path, err)
 						}
 					}
 				}

--- a/shared/idmap/idmapset_linux.go
+++ b/shared/idmap/idmapset_linux.go
@@ -607,7 +607,7 @@ func (set *IdmapSet) doUidshiftIntoContainer(dir string, testmode bool, how stri
 			return filepath.SkipDir
 		}
 
-		intUid, intGid, _, _, inode, nlink, err := shared.GetFileStat(path)
+		intUID, intGID, _, _, inode, nlink, err := shared.GetFileStat(path)
 		if err != nil {
 			return err
 		}
@@ -623,8 +623,8 @@ func (set *IdmapSet) doUidshiftIntoContainer(dir string, testmode bool, how stri
 			hardLinks = append(hardLinks, inode)
 		}
 
-		uid := int64(intUid)
-		gid := int64(intGid)
+		uid := int64(intUID)
+		gid := int64(intGID)
 		caps := []byte{}
 
 		var newuid, newgid int64
@@ -661,13 +661,13 @@ func (set *IdmapSet) doUidshiftIntoContainer(dir string, testmode bool, how stri
 
 				// Shift capabilities
 				if len(caps) != 0 {
-					rootUid := int64(0)
+					rootUID := int64(0)
 					if how == "in" {
-						rootUid, _ = set.ShiftIntoNs(0, 0)
+						rootUID, _ = set.ShiftIntoNs(0, 0)
 					}
 
 					if how != "in" || atomic.LoadInt32(&VFS3Fscaps) == VFS3FscapsSupported {
-						err = SetCaps(path, caps, rootUid)
+						err = SetCaps(path, caps, rootUID)
 						if err != nil {
 							logger.Warnf("Unable to set file capabilities on %q: %v", path, err)
 						}

--- a/shared/util_linux.go
+++ b/shared/util_linux.go
@@ -16,6 +16,7 @@ import (
 	"sync/atomic"
 	"unsafe"
 
+	"github.com/pkg/xattr"
 	"golang.org/x/sys/unix"
 
 	"github.com/lxc/lxd/lxd/revert"
@@ -68,109 +69,27 @@ func SetSize(fd int, width int, height int) (err error) {
 	return nil
 }
 
-// This uses ssize_t llistxattr(const char *path, char *list, size_t size); to
-// handle symbolic links (should it in the future be possible to set extended
-// attributed on symlinks): If path is a symbolic link the extended attributes
-// associated with the link itself are retrieved.
-func llistxattr(path string, list []byte) (sz int, err error) {
-	var _p0 *byte
-	_p0, err = unix.BytePtrFromString(path)
+// GetAllXattr retrieves all extended attributes associated with a file, directory or symbolic link.
+func GetAllXattr(path string) (map[string]string, error) {
+	xattrNames, err := xattr.LList(path)
 	if err != nil {
-		return
-	}
-
-	var _p1 unsafe.Pointer
-	if len(list) > 0 {
-		_p1 = unsafe.Pointer(&list[0])
-	} else {
-		_p1 = unsafe.Pointer(nil)
-	}
-
-	r0, _, e1 := unix.Syscall(unix.SYS_LLISTXATTR, uintptr(unsafe.Pointer(_p0)), uintptr(_p1), uintptr(len(list)))
-	sz = int(r0)
-	if e1 != 0 {
-		err = e1
-	}
-
-	return
-}
-
-// GetAllXattr retrieves all extended attributes associated with a file,
-// directory or symbolic link.
-func GetAllXattr(path string) (xattrs map[string]string, err error) {
-	// Call llistxattr() twice: First, to determine the size of the buffer
-	// we need to allocate to store the extended attributes, second, to
-	// actually store the extended attributes in the buffer. Also, check if
-	// the size/number of extended attributes hasn't increased between the
-	// two calls.
-	pre, err := llistxattr(path, nil)
-	if err != nil || pre < 0 {
-		if err == unix.EOPNOTSUPP {
+		// Some filesystems don't support llistxattr() for various reasons.
+		// Interpret this as a set of no xattrs, instead of an error.
+		if errors.Is(err, unix.EOPNOTSUPP) {
 			return nil, nil
 		}
 
-		return nil, err
+		return nil, fmt.Errorf("Failed getting extended attributes from %q: %w", path, err)
 	}
 
-	if pre == 0 {
-		return nil, nil
-	}
-
-	dest := make([]byte, pre)
-
-	post, err := llistxattr(path, dest)
-	if err != nil || post < 0 {
-		return nil, err
-	}
-
-	if post > pre {
-		return nil, fmt.Errorf("Extended attribute list size increased from %d to %d during retrieval", pre, post)
-	} else if post == 0 {
-		// fs may have filtered out all xattrs during the second call
-		return nil, nil
-	}
-
-	split := strings.Split(string(dest), "\x00")
-	if split == nil {
-		return nil, fmt.Errorf("No valid extended attribute key found")
-	}
-	// *listxattr functions return a list of  names  as  an unordered array
-	// of null-terminated character strings (attribute names are separated
-	// by null bytes ('\0')), like this: user.name1\0system.name1\0user.name2\0
-	// Since we split at the '\0'-byte the last element of the slice will be
-	// the empty string. We remove it:
-	if split[len(split)-1] == "" {
-		split = split[:len(split)-1]
-	}
-
-	xattrs = make(map[string]string, len(split))
-
-	for _, x := range split {
-		xattr := string(x)
-		// Call Getxattr() twice: First, to determine the size of the
-		// buffer we need to allocate to store the extended attributes,
-		// second, to actually store the extended attributes in the
-		// buffer. Also, check if the size of the extended attribute
-		// hasn't increased between the two calls.
-		pre, err = unix.Getxattr(path, xattr, nil)
-		if err != nil || pre < 0 {
-			return nil, err
+	var xattrs = make(map[string]string, len(xattrNames))
+	for _, xattrName := range xattrNames {
+		value, err := xattr.LGet(path, xattrName)
+		if err != nil {
+			return nil, fmt.Errorf("Failed getting %q extended attribute from %q: %w", xattrName, path, err)
 		}
 
-		dest = make([]byte, pre)
-		post := 0
-		if pre > 0 {
-			post, err = unix.Getxattr(path, xattr, dest)
-			if err != nil || post < 0 {
-				return nil, err
-			}
-		}
-
-		if post > pre {
-			return nil, fmt.Errorf("Extended attribute '%s' size increased from %d to %d during retrieval", xattr, pre, post)
-		}
-
-		xattrs[xattr] = string(dest)
+		xattrs[xattrName] = string(value)
 	}
 
 	return xattrs, nil

